blockedonweibo/weibo.py

```diff
@@ -319,17 +328,52 @@ def run(keywords,
         elif verbose=="some" and (count%10==0 or count==0):
             print(r.Index,keyword_encoded, result)
+
+        if get_canonical and result == "censored":
+            if verbose=="some" or verbose=="all":
+                print("Found censored search phrase; determining canonical censored keyword set")
+            sleep_recursive = sleep_secs if sleep is True else 0
+            potential_kws = split_search_query(keyword_encoded, cookies, sleep_recursive, res_rtn=[], known_blocked=True)
+            if verbose=="all":
+                print(potential_kws)
+
+            for kw in potential_kws:
+                test_list = [kw[:i] + kw[i + 1:] for i in range(len(kw))]
+                min_str = ""
+                for i in range(len(test_list)):
+                    if verbose=="all":
+                        print("Testing %d of %d: omitting character %s" %(i+1, len(test_list), kw[i]))
+                    if sleep_recursive:
+                        time.sleep(random.randint(math.ceil(sleep * .90), math.ceil(sleep * 1.10)))
+                    if has_censorship(test_list[i], cookies)[0] != "censored":
+                        min_str += (kw[i])
+                result_min_str, num_results_min_str = has_censorship(min_str, cookies)
+                if result_min_str == "censored":  # minStr found properly
+                    if verbose=="all" or verbose=="some":
+                        print("the minimum phrase from '%s' is: '%s'" % (kw, min_str))
+                    if insert:
+                        insert_into_database(len(sqlite_to_df(sqlite_file)), min_str, date=date, result=result_min_str,
+                                             source=r.source, num_results=num_results_min_str, notes=r.notes,
+                                             sqlite_file=sqlite_file, test_number=test_number, is_canonical=True,
+                                             orig_keyword=keyword_encoded)
+                else:
+                    print("Failed to find canonical phrase")
+
         if insert:
-            insert_into_database(len(sqlite_to_df(sqlite_file)),keyword_encoded,date=date,result=result,source=r.source,num_results=num_results,notes=r.notes,sqlite_file=sqlite_file,test_number=test_number)
+            insert_into_database(len(sqlite_to_df(sqlite_file)), keyword_encoded, date=date, result=result, source=r.source,
+                                 num_results=num_results, notes=r.notes, sqlite_file=sqlite_file, test_number=test_number,
+                                 is_canonical=False)
         if return_df:
             results_df = pd.concat([results_df,
                                     pd.DataFrame([{"date":date,
                                                    "datetime":datetime.now(),
-                                                   "keyword":r.keyword,
+                                                   "keyword":min_str if min_str is not None else r.keyword,
```

Comment:
Looks like min_str isn't in scope. Maybe move the declare on line 342 to outside the loop? Perhaps that'll fix it? (Sorry, I should probably write tests at some point to save folks time with catching stuff like this.)

```
>>> import weibo
>>> import pandas as pd
>>> sqlite_file = 'results.sqlite'
>>> weibo.create_database(sqlite_file, overwrite=True)
>>> sample_keywords_df = pd.DataFrame(
...     [{'keyword':'hello','source':'my dataframe'},
...      {'keyword':'lxb','source':'my dataframe'},
...      {'keyword':u'习胞子','source':'my dataframe'}
...     ])
>>> weibo.run(sample_keywords_df,insert=False,return_df=True)
(0, u'hello', 'has_results')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "weibo.py", line 370, in run
    "keyword":min_str if min_str is not None else r.keyword,
UnboundLocalError: local variable 'min_str' referenced before assignment
```

Reply:
Whoa.. good catch, thank you! Corrected by declaring min_str before loop